### PR TITLE
🐛 Source Salesforce: make paging work, if `cursor_field` is not changed inside one page

### LINKED ISSUE #17148: Salesforce source: Stuck in Jobs InProgress
## Environment
- **Airbyte version**: 0.40.6 - **OS Version / Instance**: AWS EC2
- **Deployment**: Kubernetes
- **Source Connector and version**: 1.0.17
- **Source Destination and version**: 0.4.35
- **Step where error happened**: Update connector

## Current Behavior
Stuck continuously logging:
> Sleeping 3.4809579870417284 seconds while waiting for Job: Opportunity/7508e000000z92DAAQ to complete. Current state: InProgress

When synchronizing the data for some fields (like Opportunity) it gets stuck in infinite job operations like the one above.
Even though the job is always new it never advances from that operation for that field.

## Expected Behavior
Read data from the opportunity field

## Steps to Reproduce
1. Activate the Opportunity field
2. Sync data

## Are you willing to submit a PR?
Yes

### PR DESCRIPTION
**Pull request by grubberr**
## What
This PR also related to this oncall https://github.com/airbytehq/oncall/issues/735

## How

Data:
--------------------------------
cursor_field=2022-10-01, Pk=1
cursor_field=2022-10-01, Pk=3
cursor_field=2022-10-01, Pk=5
cursor_field=2022-10-02, Pk=2
cursor_field=2022-10-02, Pk=4
cursor_field=2022-10-02, Pk=6

New querying with page_size=2
----------------------------------
`SELECT * FROM table WHERE cursor_field >= 2022-10-01 ORDER BY cursor_field, pk LIMIT 2`
 
  -> cursor_field=2022-10-01, Pk=1
  ->  cursor_field=2022-10-01, Pk=3

`SELECT * FROM table WHERE (cursor_field == 2022-10-01 AND pk > 3) OR (cursor_field > 2022-10-01) ORDER BY cursor_field, pk LIMIT 2`

  -> cursor_field=2022-10-01, Pk=5
  -> cursor_field=2022-10-02, Pk=2

`SELECT * FROM table WHERE cursor_field >= 2022-10-02 ORDER BY cursor_field, pk LIMIT 2`
  
   -> cursor_field=2022-10-02, Pk=2 (duplicate, it was before this PR too)
   -> cursor_field=2022-10-02, Pk=4

`SELECT * FROM table WHERE (cursor_field == 2022-10-02 AND pk > 4) OR (cursor_field > 2022-10-02) ORDER BY cursor_field, pk LIMIT 2`

  -> cursor_field=2022-10-02, Pk=6

## Recommended reading order
1. `x.java`
2. `y.python`

## 🚨 User Impact 🚨
Are there any breaking changes? What is the end result perceived by the user? If yes, please merge this PR with the 🚨🚨 emoji so changelog authors can further highlight this if needed.

## Pre-merge Checklist
Expand the relevant checklist and delete the others.

<details><summary><strong>Updating a connector</strong></summary>

### Community member or Airbyter

- [x] Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [x] Secrets in the connector's spec are annotated with `airbyte_secret`
- [x] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [x] Code reviews completed
- [x] Documentation updated
    - [ ] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [x] Changelog updated in `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.com/contributing-to-airbyte/issues-and-pull-requests)

### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items.

- [x] Create a non-forked branch based on this PR and test the below items on it
- [x] Build is successful
- [x] If new credentials are required for use in CI, add them to GSM. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci).
- [x] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing
- [ ] New Connector version released on Dockerhub and connector version bumped by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)

</details>

## Tests

<details><summary><strong>Unit</strong></summary>

*Put your unit tests output here.*

</details>

<details><summary><strong>Integration</strong></summary>

*Put your integration tests output here.*

</details>

<details><summary><strong>Acceptance</strong></summary>

*Put your acceptance tests output here.*

</details>

### CODE REVIEW

File: airbyte-integrations/connectors/source-salesforce/source_salesforce/streams.py [hunk 1]
```diff
@@ -486,8 +489,15 @@ def request_params(
...
+            if next_page_token:
+                primary_key = next_page_token.get("primary_key")
+                if primary_key and self.name not in UNSUPPORTED_FILTERING_STREAMS:
+                    query += f"AND {self.primary_key} > '{primary_key}' "
         if self.name not in UNSUPPORTED_FILTERING_STREAMS:
-            query += f"ORDER BY {self.cursor_field} ASC LIMIT {self.page_size}"
+            order_by_fields = [self.cursor_field]
+            if self.primary_key:
+                order_by_fields.append(self.primary_key)
+            query += f"ORDER BY {','.join(order_by_fields)} ASC LIMIT {self.page_size}"
```
Comment (anamargaridarl) [inline airbyte-integrations/connectors/source-salesforce/source_salesforce/streams.py:500]: Could there be a case where:
date=2 primarykey=2
date=3 primarykey=1
If so, wouldn't we miss the second entry?

Comment (grubberr) [reply airbyte-integrations/connectors/source-salesforce/source_salesforce/streams.py:500]: yes it seems you are right 

it seems we need only add primarykey to filtering if we inside that same date batch

Comment (grubberr) [reply airbyte-integrations/connectors/source-salesforce/source_salesforce/streams.py:500]: OMG things become even more complex

Comment (anamargaridarl) [reply airbyte-integrations/connectors/source-salesforce/source_salesforce/streams.py:502]: Yes, because the order by date is mandatory to make the incremental process work.

**Review by anamargaridarl — COMMENTED** [review]
Everything looks good to me now!

File: airbyte-integrations/connectors/source-salesforce/source_salesforce/streams.py [hunk 2]
```diff
@@ -481,13 +485,21 @@ def request_params(
...
         stream_date = stream_state.get(self.cursor_field)
         next_token = (next_page_token or {}).get("next_token")
+        primary_key = (next_page_token or {}).get("primary_key")
         start_date = next_token or stream_date or self.start_date
+        self.prev_start_date = start_date
 
         query = f"SELECT {','.join(selected_properties.keys())} FROM {self.name} "
         if start_date:
-            query += f"WHERE {self.cursor_field} >= {start_date} "
+            if primary_key and self.name not in UNSUPPORTED_FILTERING_STREAMS:
```
Comment (brianjlai) [inline airbyte-integrations/connectors/source-salesforce/source_salesforce/streams.py:494]: I understand why we need to check `primary_key` here, but what is the purpose of `self.name not in UNSUPPORTED_FILTERING_STREAMS` as part of the condition? Is this intended to dictate whether or not the stream supports using a WHERE clause in query?

This block feels very confusing to read with a lot of repeated conditions which makes it a little hard to reason about and I'm trying to think of ways we can streamline this query builder a bit

Comment (grubberr) [reply airbyte-integrations/connectors/source-salesforce/source_salesforce/streams.py:494]: Yes, good point. I agree it's not very obvious.

In really I just try to be consistent with this part of code, look at master branch:

https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/connectors/source-salesforce/source_salesforce/streams.py#L489

Because my "improved" query strictly relaies on `ORDER BY`
if `ORDER BY` is not used I also skip my "improved" query

Comment (anamargaridarl) [reply airbyte-integrations/connectors/source-salesforce/source_salesforce/streams.py:494]: @brianjlai Searching the documentation, some of those UNSUPPORTED_FILTERING_STREAMS fields support only the "where" filter which is consistent with the master logic. Though, I am not sure about it (since for those fields without the "order" the next date can be random ?).

I tried to find any refactors to the code to maintain that consistency but failed to do so.

**Comment by alafanechere** [comment]
@grubberr feel free to ask for  a review from my side once you have implemented @brianjlai and @anamargaridarl suggestions.

File: airbyte-integrations/connectors/source-salesforce/source_salesforce/streams.py (hunk 2 above)
Comment (alafanechere) [reply airbyte-integrations/connectors/source-salesforce/source_salesforce/streams.py:494]: > I tried to find any refactors to the code to maintain that consistency but failed to do so.

We could add a `supports_filtering` attribute to `BulkIncrementalSalesforceStream`  and set this attribute on instantiation after checking `if name in UNSUPPORTED_FILTERING_STREAMS`

File: airbyte-integrations/connectors/source-salesforce/source_salesforce/streams.py [hunk 3]
```diff
@@ -481,13 +485,21 @@ def request_params(
...
+                query += f"WHERE ({self.cursor_field} = {start_date} AND {self.primary_key} > '{primary_key}') OR ({self.cursor_field} > {start_date}) "
+            else:
+                query += f"WHERE {self.cursor_field} >= {start_date} "
```
Comment (alafanechere) [inline airbyte-integrations/connectors/source-salesforce/source_salesforce/streams.py:497]: nit 
```suggestion
           query += f"WHERE {self.cursor_field} >= {start_date} "
            if primary_key and self.name not in UNSUPPORTED_FILTERING_STREAMS:
                query += f" AND {self.primary_key} > '{primary_key}') OR ({self.cursor_field} > {start_date}) "
```

File: airbyte-integrations/connectors/source-salesforce/source_salesforce/streams.py [hunk 4]
```diff
@@ -481,13 +485,21 @@ def request_params(
...
         if self.name not in UNSUPPORTED_FILTERING_STREAMS:
-            query += f"ORDER BY {self.cursor_field} ASC LIMIT {self.page_size}"
+            order_by_fields = [self.cursor_field]
+            if self.primary_key:
+                order_by_fields.append(self.primary_key)
```
Comment (alafanechere) [inline airbyte-integrations/connectors/source-salesforce/source_salesforce/streams.py:501]: nit
```suggestion
            order_by_fields = [self.cursor_field, self.primary_key] if self.primary_key else [self.cursor_field]
```

File: airbyte-integrations/connectors/source-salesforce/source_salesforce/streams.py (hunk 1 above)
Comment (alafanechere) [reply airbyte-integrations/connectors/source-salesforce/source_salesforce/streams.py:502]: I'm under the impression this discussion is resolved with the upstream logic. Isn't it?

File: airbyte-integrations/connectors/source-salesforce/unit_tests/api_test.py [hunk 5]
```diff
@@ -528,3 +528,59 @@ def test_convert_to_standard_instance(stream_config, stream_api):
     bulk_stream = generate_stream("Account", stream_config, stream_api)
     rest_stream = bulk_stream.get_standard_instance()
     assert isinstance(rest_stream, IncrementalSalesforceStream)
+
+
+def test_bulk_stream_paging(stream_config, stream_api_pk):
+    LastModifiedDate1 = "2022-10-01T00:00:00Z"
```
Comment (alafanechere) [inline airbyte-integrations/connectors/source-salesforce/unit_tests/api_test.py:534]: Please do not capitalize variables and use snake case for non-class variables. We might have a stricter linter in the future that will ask you to change this :D

File: airbyte-integrations/connectors/source-salesforce/unit_tests/api_test.py [hunk 6]
```diff
@@ -528,3 +528,59 @@ def test_convert_to_standard_instance(stream_config, stream_api):
...
+        m.register_uri("POST", stream.path(), post_responses)
+
+        records = list(stream.read_records(sync_mode=SyncMode.full_refresh))
+
+        assert records == [
+            {"Field1": "test", "Id": 1, "LastModifiedDate": LastModifiedDate1},
+            {"Field1": "test", "Id": 3, "LastModifiedDate": LastModifiedDate1},
+            {"Field1": "test", "Id": 5, "LastModifiedDate": LastModifiedDate1},
+            {"Field1": "test", "Id": 2, "LastModifiedDate": LastModifiedDate2},
+            {"Field1": "test", "Id": 2, "LastModifiedDate": LastModifiedDate2},  # duplicate record
```
Comment (alafanechere) [inline airbyte-integrations/connectors/source-salesforce/unit_tests/api_test.py:567]: Why did you add a duplicate to the test?

File: airbyte-integrations/connectors/source-salesforce/unit_tests/api_test.py [hunk 7]
```diff
@@ -528,3 +528,59 @@ def test_convert_to_standard_instance(stream_config, stream_api):
...
+
+    csv_header = "Field1,LastModifiedDate,Id"
+    pages = [
+        [f"test,{LastModifiedDate1},1", f"test,{LastModifiedDate1},3"],
+        [f"test,{LastModifiedDate1},5", f"test,{LastModifiedDate2},2"],
+        [f"test,{LastModifiedDate2},2", f"test,{LastModifiedDate2},4"],
+        [f"test,{LastModifiedDate2},6"],
+    ]
+
+    with requests_mock.Mocker() as m:
```
Comment (alafanechere) [inline airbyte-integrations/connectors/source-salesforce/unit_tests/api_test.py:550]:
```suggestion
    with requests_mock.Mocker() as mocked_requests:
```

File: airbyte-integrations/connectors/source-salesforce/unit_tests/api_test.py [hunk 8]
```diff
@@ -528,3 +528,59 @@ def test_convert_to_standard_instance(stream_config, stream_api):
...
+            {"Field1": "test", "Id": 3, "LastModifiedDate": LastModifiedDate1},
+            {"Field1": "test", "Id": 5, "LastModifiedDate": LastModifiedDate1},
+            {"Field1": "test", "Id": 2, "LastModifiedDate": LastModifiedDate2},
+            {"Field1": "test", "Id": 2, "LastModifiedDate": LastModifiedDate2},  # duplicate record
+            {"Field1": "test", "Id": 4, "LastModifiedDate": LastModifiedDate2},
+            {"Field1": "test", "Id": 6, "LastModifiedDate": LastModifiedDate2},
+        ]
+
+        def req(i):
+            return m.request_history[i].json()["query"]
```
Comment (alafanechere) [inline airbyte-integrations/connectors/source-salesforce/unit_tests/api_test.py:573]:
```suggestion
        def get_query(mocked_requests, request_index):
            return mmocked_requests.request_history[i].json()["query"]
```

File: airbyte-integrations/connectors/source-salesforce/unit_tests/conftest.py [hunk 9]
```diff
@@ -99,6 +99,12 @@ def stream_api_v2(stream_config):
     return _stream_api(stream_config, describe_response_data=describe_response_data)
 
 
+@pytest.fixture(scope="module")
```
Comment (alafanechere) [inline airbyte-integrations/connectors/source-salesforce/unit_tests/conftest.py:102]: I suggest this fixture to be decalred in the `api_test.py` module since it's only used there. And I'm not sure the module scope is require

**Review by alafanechere — APPROVED** [review]
Nice job, thank you! Minor suggestions

File: airbyte-integrations/connectors/source-salesforce/unit_tests/api_test.py (hunk 6 above)
Comment (grubberr) [reply airbyte-integrations/connectors/source-salesforce/unit_tests/api_test.py:567]: I just want to show here that duplicates are possible
it's not related to my changes

Comment (alafanechere) [reply airbyte-integrations/connectors/source-salesforce/unit_tests/api_test.py:567]: All right, is it something you've already observed with the API?

File: airbyte-integrations/connectors/source-salesforce/unit_tests/api_test.py (hunk 5 above)
Comment (grubberr) [reply airbyte-integrations/connectors/source-salesforce/unit_tests/api_test.py]: fixed

File: airbyte-integrations/connectors/source-salesforce/unit_tests/api_test.py (hunk 8 above)
Comment (grubberr) [reply airbyte-integrations/connectors/source-salesforce/unit_tests/api_test.py]: changed

File: airbyte-integrations/connectors/source-salesforce/unit_tests/api_test.py (hunk 7 above)
Comment (grubberr) [reply airbyte-integrations/connectors/source-salesforce/unit_tests/api_test.py]: changed!

File: airbyte-integrations/connectors/source-salesforce/unit_tests/conftest.py (hunk 9 above)
Comment (grubberr) [reply airbyte-integrations/connectors/source-salesforce/unit_tests/conftest.py:102]: I just wanted to be consistent with fixture `stream_api_v2` which is also used only in `api_test.py`

File: airbyte-integrations/connectors/source-salesforce/source_salesforce/streams.py (hunk 3 above)
Comment (grubberr) [reply airbyte-integrations/connectors/source-salesforce/source_salesforce/streams.py:497]: no, it cannot be changed such way:

first query starts:
```
{self.cursor_field} = {start_date}
```

2-nd query starts:
```
{self.cursor_field} >= {start_date}
```

File: airbyte-integrations/connectors/source-salesforce/source_salesforce/streams.py (hunk 4 above)
Comment (grubberr) [reply airbyte-integrations/connectors/source-salesforce/source_salesforce/streams.py]: changed

File: airbyte-integrations/connectors/source-salesforce/unit_tests/api_test.py (hunk 6 above)
Comment (grubberr) [reply airbyte-integrations/connectors/source-salesforce/unit_tests/api_test.py]: it's possible because SF can store multiple records for the same `cursor field`.
Because we query using `>=` on subsequent queries we can grab duplicates

I mean 2-nd query can produce some result which was on 1-st query

Comment (alafanechere) [reply airbyte-integrations/connectors/source-salesforce/unit_tests/api_test.py]: Understood, can we deduplicate downstream in the connector or is it too complex? Do you mind opening an issue if you think it's something we can address later?

**Comment by grubberr** [comment]
@alafanechere what you do think can I publish and merge this PR ?
